File: server/java/src/main/java/com/onlinepayments/example/api/validators/payment/PaymentValidator.java
package com.onlinepayments.example.api.validators.payment;

import com.onlinepayments.example.api.exceptions.ApiValidationException;
import com.onlinepayments.example.api.models.payments.CreatePaymentRequest;
import com.onlinepayments.example.api.validators.BaseValidator;
import com.onlinepayments.example.api.validators.common.card.CardValidator;
import com.onlinepayments.example.api.validators.common.mandate.MandateValidator;
import com.onlinepayments.example.api.validators.common.zip.ZipValidator;
import com.onlinepayments.example.service.domain.payments.Card;
import com.onlinepayments.example.service.domain.payments.enums.PaymentMethodType;
import org.springframework.stereotype.Component;
import java.math.BigDecimal;

@Component
public class PaymentValidator extends BaseValidator {

    public void validate(CreatePaymentRequest request) {
        if (request == null) {
            throw new ApiValidationException("Request cannot be null.");
        }

        if (request.amount() == null) {
            throw new ApiValidationException("The Amount field is required.");
        }

        if (request.amount().compareTo(BigDecimal.ZERO) <= 0) {
            throw new ApiValidationException("The Amount field must be greater than zero.");
        }

        if (request.currency() == null) {
            throw new ApiValidationException("The Currency field is required.");
        }

        if (request.method() == null) {
            throw new ApiValidationException("The Method field is required.");
        }

        if (request.hostedTokenizationId() != null && request.card() != null) {
            if (!isCardEmpty(request.card())) {
                throw new ApiValidationException("If hosted tokenization id is provided, card details must not be filled.");
            }
        }

        if (request.method() == PaymentMethodType.CREDIT_CARD) {
            CardValidator.validateCard(request.card());
        }

        if (request.method() == PaymentMethodType.DIRECT_DEBIT) {
            MandateValidator.validateMandate(request.mandate());
            ZipValidator.validateZip(request.mandate().getAddress());
        }
    }

    private boolean isCardEmpty(Card card) {
        return card == null
                || (isNullOrEmpty(card.number())
                && isNullOrEmpty(card.holderName())
                && isNullOrEmpty(card.verificationCode())
                && isNullOrEmpty(card.expiryMonth())
                && isNullOrEmpty(card.expiryYear()));
    }
}

File: server/nodejs/src/presentation/validators/payment/payment-validator.ts
import { CreatePaymentRequest } from '../../models/payment/create-payment/create-payment-request';
import { ValidationError } from '../../../business/errors/validation-error';
import { MandateRules } from '../rules/mandate-rules';
import { PaymentMethodType } from '../../../business/domain/enums/payments/payment-method-type';
import { CardRules } from '../rules/card-rules';
import { Currency } from '../../../business/domain/common/enums/currency';
import { EnumUtils } from '../../utils/enum-utils';
import { AddressRules } from '../rules/address-rules';

export class PaymentValidator {
    static validate(request: CreatePaymentRequest): void {
        const errors: string[] = [];

        if (request.hostedTokenizationId && request.hostedTokenizationId.trim() !== '') {
            if (request.card) {
                const cardEmpty =
                    (!request.card.number || request.card.number.trim() === '') &&
                    (!request.card.holderName || request.card.holderName.trim() === '') &&
                    (!request.card.verificationCode || request.card.verificationCode.trim() === '') &&
                    (!request.card.expiryMonth || request.card.expiryMonth.trim() === '') &&
                    (!request.card.expiryYear || request.card.expiryYear.trim() === '');

                if (!cardEmpty) {
                    errors.push('If hosted tokenization id is provided, card details must not be filled.');
                }
            }
        }

        if (request.amount === undefined || request.amount === null) {
            errors.push('The Amount field is required.');
        } else if (!Number.isFinite(Number(request.amount))) {
            errors.push('The Amount field must be a number.');
        } else if (Number(request.amount) <= 0) {
            errors.push('The Amount field must be greater than zero.');
        }

        if (!request.currency) {
            errors.push('The Currency field is required.');
        } else if (!EnumUtils.toEnum(Currency, request.currency)) {
            errors.push('The Currency field is invalid.');
        }

        if (!request.method) {
            errors.push('The Method field is required.');
        } else if (!EnumUtils.toEnum(PaymentMethodType, request.method)) {
            errors.push('The Method field is invalid.');
        }

        if (EnumUtils.isInAllowedValues(PaymentMethodType, request.method, [PaymentMethodType.CREDIT_CARD])) {
            const cardErrors = CardRules.validateCard(request.card);
            errors.push(...cardErrors);
        }

        if (EnumUtils.isInAllowedValues(PaymentMethodType, request.method, [PaymentMethodType.DIRECT_DEBIT])) {
            const mandateErrors = MandateRules.validateMandate(request.mandate);
            errors.push(...mandateErrors);
        }

        if (
            request.billingAddress &&
            Object.values(request.billingAddress).some((v) => v != null && String(v).trim() !== '')
        ) {
            errors.push(...AddressRules.validateAddress(request.billingAddress, 'BillingAddress'));
        }

        if (
            request.shippingAddress &&
            Object.values(request.shippingAddress).some((v) => v != null && String(v).trim() !== '')
        ) {
            errors.push(...AddressRules.validateAddress(request.shippingAddress, 'ShippingAddress'));
        }

        if (errors.length > 0) {
            throw new ValidationError(errors);
        }
    }
}
